### PR TITLE
Add support for abbreviations

### PR DESCRIPTION
**Pull request by greglandrum**
This adds support for abbreviations and linker groups in two separate ways:
1. replace the atoms involved in the abbreviation/linker with a dummy atom with an abbreviation label
2. add a SUP SubstanceGroup (can be written to/read from mol blocks)

Here's an example of a molecule before and after replacement of abbreviations:
image

Other things this PR adds/changes to better support abbreviations

1. two new atomic properties that can control the label used when rendering an atom: `_displayLabel` and `_displayLabelW`. These take priority over `atomLabel` and can be used by other code in the future any time it's better when the label is displayed differently depending on whether a bond is coming in from the left or right. These, along with the `atomLabel` property are now also explained in the getting started document

### CODE REVIEW

File: Code/GraphMol/Abbreviations/Abbreviations.cpp [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+std::vector<AbbreviationMatch> findApplicableAbbreviationMatches(
+    const ROMol& mol, const std::vector<AbbreviationDefinition>& abbrevs,
+    double maxCoverage) {
+  std::vector<AbbreviationMatch> res;
+  auto nAtoms = mol.getNumAtoms();
+  if (!nAtoms || abbrevs.empty()) {
+    return res;
+  }
+
+  MolOps::fastFindRings(mol);
```
Comment (bp-kelley) [inline Code/GraphMol/Abbreviations/Abbreviations.cpp:114]: Haven't we had issues with fast find rings having issues down stream?  If we call fast find rings, should we remove the rings at the end?

Comment (bp-kelley) [reply Code/GraphMol/Abbreviations/Abbreviations.cpp:114]: (This might be on a copy)

File: Code/GraphMol/Abbreviations/Abbreviations.cpp [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+    CHECK_INVARIANT(abbrev.mol, "molecule is null");
+    if (maxCoverage > 0) {
+      unsigned int nDummies;
+      abbrev.mol->getProp(common_properties::numDummies, nDummies);
+      if (double(abbrev.mol->getNumAtoms() - nDummies) / nAtoms >=
+          maxCoverage) {
+        continue;
+      }
+    }
+    auto matches = SubstructMatch(mol, *abbrev.mol);
```
Comment (bp-kelley) [inline Code/GraphMol/Abbreviations/Abbreviations.cpp:131]: What happens with overlapping matches or linkers?

File: Code/GraphMol/Abbreviations/Abbreviations.cpp (hunk 1 above)
Comment (greglandrum) [reply Code/GraphMol/Abbreviations/Abbreviations.cpp:114]: It's not on a copy, but FastFindRings() will only do something on molecules that don't already have ring info (i.e. unsanitized molecules), so I don't think it's a problem.

File: Code/GraphMol/Abbreviations/Abbreviations.cpp (hunk 2 above)
Comment (greglandrum) [reply Code/GraphMol/Abbreviations/Abbreviations.cpp:131]: only the first one is used. That's the loop at line 140

File: Code/GraphMol/Abbreviations/Abbreviations.cpp (hunk 1 above)
Comment (bp-kelley) [reply Code/GraphMol/Abbreviations/Abbreviations.cpp:114]: For downstream effects, I was referring to things like this:

https://github.com/rdkit/rdkit/issues/1021

Where GetSymmSSSR is what the sanitizer would call instead.

Comment (greglandrum) [reply Code/GraphMol/Abbreviations/Abbreviations.cpp:114]: Yeah, I appreciate the problem. One should really sanitize things before using them, but people won’t. 
I’ll close this

Comment (greglandrum) [reply Code/GraphMol/Abbreviations/Abbreviations.cpp:114]: Yeah, I appreciate the problem. One should really sanitize things before using them, but people won’t. 
I’ll close this

**Pull request merged**